feat: introduce parameter for returning full variable values

kristinkomschow opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->
- added new `truncateValues` boolean parameter to user task variable and variables search requests
- added test coverage for returning full variables on those search requests

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

relates to https://github.com/camunda/camunda/issues/39189

berkaycanbc committed:
feat: introduce parameter for returning full variable values

relates to https://github.com/camunda/camunda/issues/39189

kristinkomschow committed:
test: added test coverage for includeFullValue search parameter

tmetzke commented on clients/java/src/main/java/io/camunda/client/CamundaClient.java:
```diff
@@ -1902,6 +1918,25 @@ DecisionRequirementsGetXmlRequest newDecisionRequirementsGetXmlRequest(
+   *  @return a builder for the user task variable search request
+   */
+  UserTaskVariableSearchRequest newUserTaskVariableSearchRequest(
+      long userTaskKey, boolean truncateValues);
```
💭 So far, in the client, we only use required parameters in the method parameter list. So, keys, usually, because they are required on the URL path, for example. Since `truncateValues` is an optional parameter, it's consistent to integrate it with the builder pattern for the commands we usually have.

**How does that work?**
* Add a `UserTaskVariableSearchRequest truncateValues(boolean truncateValues)` method to the `UserTaskVariableSearchRequest` interface (and the variable search request).
* Implement that method in the related request implementation classes like `UserTaskVariableSearchRequestImpl`, storing the boolean value in the request (like you did with the constructor value).

This way, users can use this new interface method if they like and add it anywhere in the client call, e.g. like
```
client
    .newUserTaskVariableSearchRequest(userTaskKey)
    .filter(f -> f.name("foo"))
    .truncateValues(false) <-- new option
    .page(p -> p.limit(2))
    .send();
```

tmetzke reviewed (changes requested):
👍 Looks good overall, going into the right direction 🚀 
💭 I'd suggest staying consistent with how optional command parameters are handled in the Camunda client, reusing the builder pattern instead of a new client method.
🔧 Let's add Camunda client tests for this as well. Extending the acceptance tests for user task and variable search might also be a good idea to test this is considered correctly end-to-end.

tmetzke commented on clients/java/src/main/java/io/camunda/client/impl/CamundaClientImpl.java:
```diff
@@ -1135,7 +1140,14 @@ public VariableGetRequest newVariableGetRequest(final long variableKey) {
   @Override
   public UserTaskVariableSearchRequest newUserTaskVariableSearchRequest(final long userTaskKey) {
-    return new UserTaskVariableSearchRequestImpl(httpClient, jsonMapper, userTaskKey);
+    return new UserTaskVariableSearchRequestImpl(httpClient, jsonMapper, userTaskKey, false);
```
👍 Fair point, but this will go away with the new builder pattern approach.
💭 @kristinkomschow, in the request implementation classes, we can also make this a "big" Boolean instead. Only if it was set by a user in the client, we send it along then. Otherwise, we'll always send it from the client, but that's not necessary, I think.

tmetzke commented on qa/acceptance-tests/src/test/java/io/camunda/it/client/UserTaskSearchTest.java:
```diff
@@ -779,6 +763,43 @@ void shouldReturnUserTaskVariablesFilteredByIn() {
   }
 
+  @Test
+  void shouldReturnLargeUserTaskVariableTruncated() {
```
```suggestion
  void shouldReturnLargeUserTaskVariableNotTruncated() {
```

tmetzke replied:
🙃 The test method names are mixed up, looking at the test code. Not a big deal, I was just shortly confused when looking at the name and the code 😅

tmetzke reviewed (commented):
💭 Noticed some minor issue with the new boolean value and potential `null` values.

tmetzke reviewed (approved):
👍 Looks good overall, nice work 🦖 
🔧 The acceptance tests can be cleaned up a bit (using an existing helper class) and made more robust (avoiding size threshold magic numbers).

💭 Do we want to allow users to set a `null` to the `truncateValues` boolean? What's the purpose of that? Just wondering why I would use that as a client user. It's `null` by default, so I would only ever set it to `false` to get full values, I guess?
💭 Thinking about it again, we could even "decouple" this from the raw API and rather introduce a method named `withFullValues()`. That makes it more DSL-like for the client, abstracting from the boolean value to send. This could be handled transparently by the client implementation. @megglos, any thoughts on this? Can also be a follow-up but must be clarified before we backport and release with 8.8.

tmetzke commented on zeebe/gateway-protocol/src/main/proto/v2/variables.yaml:
```diff
@@ -8,7 +8,14 @@ paths:
+        - name: truncateValues
+          in: query
+          required: false
+          description: Option to return full variable values in the response. If not provided, it's considered true.
```
@kristinkomschow, I think rewording this makes sense 👍

megglos commented:
> 💭 Thinking about it again, we could even "decouple" this from the raw API and rather introduce a method named withFullValues(). That makes it more DSL-like for the client, abstracting from the boolean value to send. This could be handled transparently by the client implementation. @megglos, any thoughts on this? Can also be a follow-up but must be clarified before we backport and release with 8.8.

I agree that the name `withFullValues` makes it more DSL like, what I don't like about it, is the misalignment with the API naming.

I guess it mainly depends on how a user goes about making use of the client, whether that matters. Do they even check the API docs? I actually don't know. On the other hand it's the only reference doc we have, the client lacks comprehensive reference docs on the functionality provided. My assumption here is that users will rather explore the client in Java first and only in doubt double check the API Reference docs in case javadoc is not sufficient.

Effectively we have a hand crafted language specific API layer within the client anyway, thus I'm in favor of making it more convenient via the suggested `withFullValues()` + some javadoc explaining the default is truncation and this allows to opt-in to get fully expanded values.

kristinkomschow committed:
test: refactored user task and variable search tests

- using TestHelper methods instead of private test methods
- applying IDE suggestions to simplify `isEqualTo` assertions, joining assertions, removing unused variables

kristinkomschow commented:
Changes:
- Import DEFAULT_VARIABLE_SIZE_THRESHOLD from schema module instead of setting it in the testcases
- Changed request builder method to be opt-in `withFullValues()` and added java-doc
- Reworded API description as suggested by Copilot
- Using TestHelper methods instead of private test methods

@tmetzke @megglos 
I left the `withFullValues` change in an own commit but will squash it before merging. My reason is that I had one more thought for the question:
> Do we want to allow users to set a null to the truncateValues boolean?

The semantic difference when allowing a null value would be:
`null`: use server default, which may change at some point (unlikely, but can't fully be ruled out)
`true`: explicitly truncate values
`false`: explicitly do not truncate values

I personally like that distinction, but if you stand by the `withFullValues()` because it currently makes more sense from a DevEx perspective and there is no foreseeable reason for the server behavior to change, we can go with that.

tmetzke commented on zeebe/gateway-protocol/src/main/proto/v2/user-tasks.yaml:
```diff
@@ -289,14 +289,20 @@ paths:
+        - name: truncateValues
+          in: query
+          required: false
+          description: When true (default), long variable values in the response are truncated. When false, full variable values are returned
```
```suggestion
          description: When true (default), long variable values in the response are truncated. When false, full variable values are returned.
```

tmetzke commented on zeebe/gateway-protocol/src/main/proto/v2/variables.yaml:
```diff
@@ -8,7 +8,14 @@ paths:
+        - name: truncateValues
+          in: query
+          required: false
+          description: When true (default), long variable values in the response are truncated. When false, full variable values are returned
```
```suggestion
          description: When true (default), long variable values in the response are truncated. When false, full variable values are returned.
```

tmetzke reviewed (approved):
👍 Nice job, @kristinkomschow 🚀 
🔧 Just some very minor follow-ups, nothing blocking, so pre-approving 🍪

kristinkomschow commented:
Applied the remaining small review hints and squashed everything into sensible commits
